File: frontend/lib/models/product.dart
class Product {
  final int id;
  final String name;
  final String description;
  final double price;
  final double? promotionalPrice;
  final String? imageUrl;
  final bool isAvailable;
  final int? preparationTime;
  final int restaurantId;
  final Category? category;

  Product({
    required this.id,
    required this.name,
    required this.description,
    required this.price,
    this.promotionalPrice,
    this.imageUrl,
    required this.isAvailable,
    this.preparationTime,
    required this.restaurantId,
    this.category,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'],
      name: json['name'],
      description: json['description'] ?? '',
      price: (json['price'] as num).toDouble(),
      promotionalPrice: json['promotionalPrice']?.toDouble(),
      imageUrl: json['imageUrl'],
      isAvailable: json['isAvailable'] ?? true,
      preparationTime: json['preparationTime'],
      restaurantId: json['restaurantId'],
      category: json['category'] != null ? Category.fromJson(json['category']) : null,
    );
  }

  double get currentPrice => promotionalPrice ?? price;
  bool get hasPromotion => promotionalPrice != null && promotionalPrice! < price;
  String get formattedPrice => 'R\$ ${currentPrice.toStringAsFixed(2).replaceAll('.', ',')}';
  String get formattedOriginalPrice => 'R\$ ${price.toStringAsFixed(2).replaceAll('.', ',')}';
}

class Category {
  final int id;
  final String name;
  final String? description;
  final String? iconUrl;

  Category({
    required this.id,
    required this.name,
    this.description,
    this.iconUrl,
  });

  factory Category.fromJson(Map<String, dynamic> json) {
    return Category(
      id: json['id'],
      name: json['name'],
      description: json['description'],
      iconUrl: json['iconUrl'],
    );
  }
}

File: backend/src/main/java/com/openbag/entity/Product.java
package com.openbag.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "products")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotBlank
    @Size(max = 100)
    private String name;

    @Size(max = 500)
    private String description;

    @NotNull
    @Column(precision = 10, scale = 2)
    private BigDecimal price;

    @Column(name = "promotional_price", precision = 10, scale = 2)
    private BigDecimal promotionalPrice;

    @Column(name = "image_url")
    private String imageUrl;

    @Column(name = "is_available")
    private boolean isAvailable = true;

    @Column(name = "is_active")
    private boolean isActive = true;

    @Column(name = "preparation_time")
    private Integer preparationTime;

    @CreationTimestamp
    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "restaurant_id")
    private Restaurant restaurant;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    private Category category;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<OrderItem> orderItems = new ArrayList<>();

    public Product(String name, String description, BigDecimal price, Restaurant restaurant, Category category) {
        this.name = name;
        this.description = description;
        this.price = price;
        this.restaurant = restaurant;
        this.category = category;
    }

    public BigDecimal getCurrentPrice() {
        return promotionalPrice != null ? promotionalPrice : price;
    }
}
